### PR TITLE
enable 3 options for optimization

### PR DESCRIPTION
**Pull request by yetongumich**
Enable calling optimization with 3 options: unconstrained/penalty method/Augmented Lagrangian method. 

During implementation, several changes are made for the constrained optimizer to work with kinematics:
- rewrite PointGoalFactor with ExpressionFactor
- create EqualityConstraints class with "add" operations (same as FactorGraph)

Also add a flag to represent the option to add PointGoal as objective or constraint

### CODE REVIEW

File: gtdynamics/factors/PointGoalFactor.h [hunk 1]
```diff
@@ -16,78 +16,52 @@
...
-  PointGoalFactor(gtsam::Key pose_key,
-                  const gtsam::noiseModel::Base::shared_ptr &cost_model,
+inline gtsam::Vector3_ PointGoalConstraint(gtsam::Key pose_key,
                   const gtsam::Point3 &point_com,
-                  const gtsam::Point3 &goal_point)
-      : Base(cost_model, pose_key),
-        point_com_(point_com),
-        goal_point_(goal_point) {}
+                  const gtsam::Point3 &goal_point) {
+  // const Point3 point_world = wTcom.transformFrom(point_com_, H_pose);
```
Comment (dellaert) [inline gtdynamics/factors/PointGoalFactor.h:39]: remove?

File: gtdynamics/kinematics/KinematicsSlice.cpp [hunk 2]
```diff
@@ -105,11 +137,22 @@ Values Kinematics::initialValues<Slice>(const Slice& slice, const Robot& robot,
...
+  // Robot kinematics constraints
+  auto constraints = this->constraints(slice, robot);
+  NonlinearFactorGraph graph;
+
+  // Contact goals
+  if (contact_goals_as_constraints) {
+    constraints.add(this->pointGoalConstraints(slice, contact_goals));
+  }
+  else {
+    graph.add(pointGoalObjectives(slice, contact_goals));
```
Comment (dellaert) [inline gtdynamics/kinematics/KinematicsSlice.cpp:152]: Why not do this with constraint->createFactor?

File: gtdynamics/kinematics/KinematicsSlice.cpp [hunk 3]
```diff
@@ -105,11 +137,22 @@ Values Kinematics::initialValues<Slice>(const Slice& slice, const Robot& robot,
...
 
-  // Add objectives.
-  graph.add(pointGoalObjectives(slice, contact_goals));
+  // Robot kinematics constraints
+  auto constraints = this->constraints(slice, robot);
+  NonlinearFactorGraph graph;
+
+  // Contact goals
+  if (contact_goals_as_constraints) {
+    constraints.add(this->pointGoalConstraints(slice, contact_goals));
```
Comment (dellaert) [inline gtdynamics/kinematics/KinematicsSlice.cpp:149]: What is `EqualityConstraints` anyway? I thought we had decided on just using a vector?

File: gtdynamics/kinematics/KinematicsSlice.cpp [hunk 4]
```diff
@@ -51,6 +51,22 @@ NonlinearFactorGraph Kinematics::graph<Slice>(const Slice& slice,
...
+template <>
+EqualityConstraints Kinematics::constraints<Slice>(const Slice& slice,
+                                              const Robot& robot) const {
+  EqualityConstraints constraints;
+
+  // Constrain kinematics at joints.
+  for (auto&& joint : robot.joints()) {
+    auto constraint_expr = joint->poseConstraint(slice.k);
+    gtsam::Vector6 tolerance = p_->p_cost_model->sigmas();
+    constraints.add(VectorExpressionEquality<6>(constraint_expr, tolerance));
```
Comment (dellaert) [inline gtdynamics/kinematics/KinematicsSlice.cpp:63]: use emplace variant

File: gtdynamics/kinematics/KinematicsSlice.cpp [hunk 5]
```diff
@@ -51,6 +51,22 @@ NonlinearFactorGraph Kinematics::graph<Slice>(const Slice& slice,
...
 
+template <>
+EqualityConstraints Kinematics::constraints<Slice>(const Slice& slice,
+                                              const Robot& robot) const {
+  EqualityConstraints constraints;
+
+  // Constrain kinematics at joints.
+  for (auto&& joint : robot.joints()) {
+    auto constraint_expr = joint->poseConstraint(slice.k);
+    gtsam::Vector6 tolerance = p_->p_cost_model->sigmas();
```
Comment (dellaert) [inline gtdynamics/kinematics/KinematicsSlice.cpp:62]: cost models should now be replaced with tolerances I think. But we can do that in a different PR

File: gtdynamics/kinematics/KinematicsSlice.cpp [hunk 6]
```diff
@@ -105,11 +137,22 @@ Values Kinematics::initialValues<Slice>(const Slice& slice, const Robot& robot,
 
 template <>
 Values Kinematics::inverse<Slice>(const Slice& slice, const Robot& robot,
-                                  const ContactGoals& contact_goals) const {
-  auto graph = this->graph(slice, robot);
+                                  const ContactGoals& contact_goals,
+                                  bool contact_goals_as_constraints) const {
```
Comment (dellaert) [inline gtdynamics/kinematics/KinematicsSlice.cpp:141]: Why have this flag? Comment on this in PR comment,

File: gtdynamics/optimizer/EqualityConstraint.h [hunk 7]
```diff
@@ -135,7 +135,33 @@ class VectorExpressionEquality : public EqualityConstraint {
...
+
+  template <typename DERIVEDCONSTRAINT>
+  using IsDerived = typename std::enable_if<
+      std::is_base_of<EqualityConstraint, DERIVEDCONSTRAINT>::value>::type;
+
+ public:
+  EqualityConstraints() : Base() {}
+
+  /**
+   * Add a constraint by value, will be copy-constructed (use push_back with a
```
Comment (dellaert) [inline gtdynamics/optimizer/EqualityConstraint.h:152]: let's not make this possible?

File: gtdynamics/optimizer/EqualityConstraint.h [hunk 8]
```diff
@@ -135,7 +135,33 @@ class VectorExpressionEquality : public EqualityConstraint {
...
+   * shared_ptr to avoid the copy).
+   */
+  template <class DERIVEDCONSTRAINT>
+  IsDerived<DERIVEDCONSTRAINT> add(const DERIVEDCONSTRAINT& constraint) {
+    push_back(boost::allocate_shared<DERIVEDCONSTRAINT>(
+            Eigen::aligned_allocator<DERIVEDCONSTRAINT>(), constraint));
+  }
+
+  void add (const EqualityConstraints& other) {
```
Comment (dellaert) [inline gtdynamics/optimizer/EqualityConstraint.h:161]: is this why this class exists?

File: gtdynamics/optimizer/EqualityConstraint.h [hunk 9]
```diff
@@ -135,7 +135,33 @@ class VectorExpressionEquality : public EqualityConstraint {
 };
 
 /// Container of EqualityConstraint.
-typedef std::vector<EqualityConstraint::shared_ptr> EqualityConstraints;
+// typedef std::vector<EqualityConstraint::shared_ptr> EqualityConstraints;
+
+class EqualityConstraints : public std::vector<EqualityConstraint::shared_ptr> {
```
Comment (dellaert) [inline gtdynamics/optimizer/EqualityConstraint.h:140]: I think you should still use the typedef (but use `using`)

File: gtdynamics/optimizer/Optimizer.cpp [hunk 10]
```diff
@@ -21,9 +23,41 @@ using gtsam::Values;
 
 Values Optimizer::optimize(const NonlinearFactorGraph& graph,
                            const Values& initial_values) const {
+  std::cout<< "running LM\n";
```
Comment (dellaert) [inline gtdynamics/optimizer/Optimizer.cpp:26]: Remove in final PR. Why do we still have this optimize ?

File: gtdynamics/optimizer/Optimizer.cpp [hunk 11]
```diff
@@ -21,9 +23,41 @@ using gtsam::Values;
...
   gtsam::LevenbergMarquardtOptimizer optimizer(graph, initial_values,
                                                p_->lm_parameters);
   const Values result = optimizer.optimize();
+  std::cout << "optimization done\n";
   return result;
 }
+
+Values Optimizer::optimize(const gtsam::NonlinearFactorGraph& graph,
```
Comment (dellaert) [inline gtdynamics/optimizer/Optimizer.cpp:34]: Nice !

File: gtdynamics/optimizer/Optimizer.cpp [hunk 12]
```diff
@@ -21,9 +23,41 @@ using gtsam::Values;
...
+  if (p_->method == OptimizationParameters::Method::UNCONSTRAINED) {
+    auto merit_graph = graph;
+    for (const auto& constraint: constraints) {
+      merit_graph.add(constraint->createFactor(1.0));
+    }
+    std::cout << "constructed merit graph\n";
+    return optimize(merit_graph, initial_values);
+  }
+  else if (p_->method == OptimizationParameters::Method::PENALTY) {
+    auto params = boost::make_shared<PenaltyMethodParameters>();
```
Comment (dellaert) [inline gtdynamics/optimizer/Optimizer.cpp:47]: this still draft I guess.. Parameters need to be passed into constructor

File: gtdynamics/optimizer/Optimizer.h [hunk 13]
```diff
@@ -54,5 +62,16 @@ class Optimizer {
...
                          const gtsam::Values& initial_values) const;
+
+  /**
+   * @brief optimize with constraints using optimizer settings.
+   *
+   * @param graph a Nonlinear factor graph built by derived class
+   * @param initial_values Initial values for all variables.
+   * @return Values The result of the optimization.
+   */
+  gtsam::Values optimize(const gtsam::NonlinearFactorGraph& graph,
```
Comment (dellaert) [inline gtdynamics/optimizer/Optimizer.h:73]: Should be the only optimizer.

File: gtdynamics/utils/Trajectory.h [hunk 14]
```diff
@@ -191,7 +191,7 @@ class Trajectory {
    * @param[in] cost_model        Noise model
    * @param[in] goal_point        target goal point
    */
-  PointGoalFactor pointGoalFactor(const Robot &robot,
+  gtsam::ExpressionFactor<gtsam::Vector3> pointGoalFactor(const Robot &robot,
```
Comment (dellaert) [inline gtdynamics/utils/Trajectory.h:194]: format

File: tests/testEqualityConstraint.cpp [hunk 15]
```diff
@@ -146,6 +146,25 @@ TEST(EqualityConstraint, VectorExpressionEquality) {
   EXPECT_CORRECT_FACTOR_JACOBIANS(*merit_factor, values2, 1e-7, 1e-5);
 }
 
+TEST(EqualityConstraint, Container) {
```
Comment (dellaert) [inline tests/testEqualityConstraint.cpp:149]: Had not expected this in this PR

File: tests/testKinematicsInterval.cpp [hunk 16]
```diff
@@ -37,6 +37,7 @@ TEST(Interval, InverseKinematics) {
 
   // Instantiate kinematics algorithms
   auto parameters = boost::make_shared<KinematicsParameters>();
+  parameters->method = OptimizationParameters::Method::UNCONSTRAINED;
```
Comment (dellaert) [inline tests/testKinematicsInterval.cpp:40]: instead of UNCONSTRAINED I would call this SOFT_CONSTRAINTS

File: tests/testObjectiveFactors.cpp [hunk 17]
```diff
@@ -113,15 +114,15 @@ TEST(Phase, AddGoals) {
       SimpleSwingTrajectory(stance_point, step, num_swing_steps), id);
   EXPECT_LONGS_EQUAL(num_swing_steps, swing_factors.size());
 
-  // Last goal point should have moved just in front of stance_point
-  auto g = boost::dynamic_pointer_cast<PointGoalFactor>(swing_factors.front());
-  EXPECT(assert_equal<Point3>(stance_point + Point3(0.01, 0, 0.035588),
-                              g->goalPoint(), 1e-5));
+  // // Last goal point should have moved just in front of stance_point
```
Comment (dellaert) [inline tests/testObjectiveFactors.cpp:117]: ???

File: tests/testPhase.cpp [hunk 18]
```diff
@@ -58,7 +58,7 @@ TEST(Phase, All) {
 
   // contactPointObjectives
   const Point3 step(0, 0.4, 0);
-  const gtsam::SharedNoiseModel cost_model = nullptr;
+  const gtsam::SharedNoiseModel cost_model = gtsam::noiseModel::Unit::Create(3);
```
Comment (dellaert) [inline tests/testPhase.cpp:61]: why?

File: tests/testTrajectory.cpp [hunk 19]
```diff
@@ -105,11 +105,11 @@ TEST(Trajectory, error) {
   // steps = 2+3 per walk cycle, 5 legs involved
   const size_t expected = repeat * ((2 + 3) * 5);
   EXPECT_LONGS_EQUAL(expected, contact_link_objectives.size());
-  // regression
-  auto last_factor = boost::dynamic_pointer_cast<PointGoalFactor>(
-      contact_link_objectives.back());
-  EXPECT(gtsam::assert_equal(gtsam::Point3(-0.190001, -0.300151, 0.000151302),
-                             last_factor->goalPoint(), 1e-5));
+  // // regression
+  // auto last_factor = boost::dynamic_pointer_cast<PointGoalFactor>(
```
Comment (dellaert) [inline tests/testTrajectory.cpp:109]: ???

File: tests/testWalkCycle.cpp [hunk 20]
```diff
@@ -102,7 +102,7 @@ TEST(WalkCycle, objectives) {
   EXPECT(gtsam::assert_equal(goal_RH, cp_goals["lower3"], 1e-6));
 
   const Point3 step(0, 0.4, 0);
-  const gtsam::SharedNoiseModel cost_model = nullptr;
+  const gtsam::SharedNoiseModel cost_model = gtsam::noiseModel::Unit::Create(3);
```
Comment (dellaert) [inline tests/testWalkCycle.cpp:105]: why?

File: gtdynamics/kinematics/KinematicsInterval.cpp [hunk 21]
```diff
@@ -44,6 +54,16 @@ NonlinearFactorGraph Kinematics::pointGoalObjectives<Interval>(
   return graph;
 }
 
+template <>
+EqualityConstraints Kinematics::pointGoalConstraints<Interval>(
```
Comment (dellaert) [inline gtdynamics/kinematics/KinematicsInterval.cpp:58]: Why do we have both objectives and constraints?

File: tests/testPhase.cpp (hunk 18 above)
Comment (yetongumich) [reply tests/testPhase.cpp:61]: ExpressionFactors won't allow noiseModel to be nullptr

File: tests/testTrajectory.cpp (hunk 19 above)
Comment (yetongumich) [reply tests/testTrajectory.cpp:109]: As we rewrite the PointGoalFactor with expressions, we won't be able to access the member function goalPoint()

File: gtdynamics/kinematics/KinematicsInterval.cpp (hunk 21 above)
Comment (yetongumich) [reply gtdynamics/kinematics/KinematicsInterval.cpp:58]: Sometimes we want to have the point goals as objectives (there are some cases in which not all point goals can be reached), and sometimes we want them to be constraints. I can keep only the constraints if we don't need the objective version.

**Comment by dellaert** [comment]
Let’s aim for me having doing a second review before we meet?

File: tests/testKinematicsInterval.cpp (hunk 16 above)
Comment (yetongumich) [reply tests/testKinematicsInterval.cpp:40]: done

File: gtdynamics/factors/PointGoalFactor.h (hunk 1 above)
Comment (yetongumich) [reply gtdynamics/factors/PointGoalFactor.h:39]: done

File: tests/testObjectiveFactors.cpp (hunk 17 above)
Comment (yetongumich) [reply tests/testObjectiveFactors.cpp:117]: this is due to rewriting PointGoalFactor as Expressionfactor loses the member function `goalPoint()`. I now make PointGoalFactor inherit from ExpressionFactor, and re-added the `goalPoint()` function to make these tests still working.

File: gtdynamics/kinematics/KinematicsSlice.cpp (hunk 4 above)
Comment (yetongumich) [reply gtdynamics/kinematics/KinematicsSlice.cpp:63]: done

File: gtdynamics/optimizer/EqualityConstraint.h (hunk 9 above)
Comment (yetongumich) [reply gtdynamics/optimizer/EqualityConstraint.h:140]: done

File: gtdynamics/optimizer/Optimizer.cpp (hunk 10 above)
Comment (yetongumich) [reply gtdynamics/optimizer/Optimizer.cpp:26]: done. We keep this optimize for compatibility of existing code and test.

File: gtdynamics/utils/Trajectory.h (hunk 14 above)
Comment (yetongumich) [reply gtdynamics/utils/Trajectory.h:194]: done

File: tests/testWalkCycle.cpp (hunk 20 above)
Comment (yetongumich) [reply tests/testWalkCycle.cpp:105]: same.

File: gtdynamics/optimizer/EqualityConstraint.h (hunk 7 above)
Comment (yetongumich) [reply gtdynamics/optimizer/EqualityConstraint.h:152]: done

File: gtdynamics/optimizer/EqualityConstraint.h (hunk 8 above)
Comment (yetongumich) [reply gtdynamics/optimizer/EqualityConstraint.h:161]: Yes. And I also add the emplace_shared function (same implementation as `emplace_shared` in FactorGraph)

File: tests/testEqualityConstraint.cpp (hunk 15 above)
Comment (yetongumich) [reply tests/testEqualityConstraint.cpp:149]: Since we added EqualityConstraints class, this is the corresponding unit test

File: gtdynamics/kinematics/KinematicsSlice.cpp (hunk 3 above)
Comment (yetongumich) [reply gtdynamics/kinematics/KinematicsSlice.cpp:149]: I discovered the need of `emplace_shared` and adding all equalityconstraints from another `EqualityConstraint` class. (same as member functions of `FactorGraph`)

File: gtdynamics/optimizer/Optimizer.h (hunk 13 above)
Comment (yetongumich) [reply gtdynamics/optimizer/Optimizer.h:73]: Keep the old one for compatibility of old code.

File: gtdynamics/kinematics/KinematicsSlice.cpp (hunk 6 above)
Comment (yetongumich) [reply gtdynamics/kinematics/KinematicsSlice.cpp:141]: done

File: gtdynamics/kinematics/KinematicsSlice.cpp (hunk 5 above)
Comment (yetongumich) [reply gtdynamics/kinematics/KinematicsSlice.cpp:62]: added a TODO

File: gtdynamics/kinematics/KinematicsSlice.cpp (hunk 2 above)
Comment (yetongumich) [reply gtdynamics/kinematics/KinematicsSlice.cpp:154]: since we already have the `pointGoalObjectives` function.

File: gtdynamics/optimizer/Optimizer.cpp (hunk 12 above)
Comment (yetongumich) [reply gtdynamics/optimizer/Optimizer.cpp:44]: Done

File: gtdynamics/factors/PointGoalFactor.h [hunk 22]
```diff
@@ -101,9 +91,16 @@ class PointGoalFactor : public gtsam::NoiseModelFactor1<gtsam::Pose3> {
...
  */
-gtsam::NonlinearFactorGraph PointGoalFactors(
+inline gtsam::NonlinearFactorGraph PointGoalFactors(
     gtsam::Key first_key, const gtsam::noiseModel::Base::shared_ptr &cost_model,
     const gtsam::Point3 &point_com,
-    const std::vector<gtsam::Point3> &goal_trajectory);
+    const std::vector<gtsam::Point3> &goal_trajectory) {
+  gtsam::NonlinearFactorGraph factors;
+  for (auto &&goal_point : goal_trajectory) {
+    factors.add(PointGoalFactor(first_key, cost_model, point_com, goal_point));
```
Comment (dellaert) [inline gtdynamics/factors/PointGoalFactor.h:100]: Use emplace_shared. ˋadd` is obsolete.

File: gtdynamics/kinematics/KinematicsSlice.cpp [hunk 23]
```diff
@@ -59,13 +76,30 @@ NonlinearFactorGraph Kinematics::pointGoalObjectives<Slice>(
   // Add objectives.
   for (const ContactGoal& goal : contact_goals) {
     const gtsam::Key pose_key = internal::PoseKey(goal.link()->id(), slice.k);
-    graph.emplace_shared<PointGoalFactor>(pose_key, p_->g_cost_model,
-                                          goal.contactInCoM(), goal.goal_point);
+    graph.add(PointGoalFactor(pose_key, p_->g_cost_model,
```
Comment (dellaert) [inline gtdynamics/kinematics/KinematicsSlice.cpp:79]: No, it really should be emplace_shared, I think :-) why add an extra copy?

File: gtdynamics/kinematics/KinematicsSlice.cpp [hunk 24]
```diff
@@ -59,13 +76,30 @@ NonlinearFactorGraph Kinematics::pointGoalObjectives<Slice>(
...
+EqualityConstraints Kinematics::pointGoalConstraints<Slice>(
+    const Slice& slice, const ContactGoals& contact_goals) const {
+  EqualityConstraints constraints;
+
+  // Add objectives.
+  for (const ContactGoal& goal : contact_goals) {
+    const gtsam::Key pose_key = internal::PoseKey(goal.link()->id(), slice.k);
+    auto constraint_expr =
+        PointGoalConstraint(pose_key, goal.contactInCoM(), goal.goal_point);
+    gtsam::Vector3 tolerance = p_->g_cost_model->sigmas();
```
Comment (dellaert) [inline gtdynamics/kinematics/KinematicsSlice.cpp:96]: Out of the loop.

File: gtdynamics/kinematics/KinematicsSlice.cpp [hunk 25]
```diff
@@ -51,6 +51,23 @@ NonlinearFactorGraph Kinematics::graph<Slice>(const Slice& slice,
...
+template <>
+EqualityConstraints Kinematics::constraints<Slice>(const Slice& slice,
+                                              const Robot& robot) const {
+  EqualityConstraints constraints;
+
+  // Constrain kinematics at joints.
+  for (auto&& joint : robot.joints()) {
+    auto constraint_expr = joint->poseConstraint(slice.k);
+    gtsam::Vector6 tolerance = p_->p_cost_model->sigmas();
+    constraints.emplace_shared<VectorExpressionEquality<6>>(constraint_expr,
```
Comment (dellaert) [inline gtdynamics/kinematics/KinematicsSlice.cpp:63]: Tolerance is constant over loop: take out of the loop

File: gtdynamics/optimizer/AugmentedLagrangianOptimizer.h [hunk 26]
```diff
@@ -20,11 +20,15 @@ namespace gtdynamics {
...
 
-  AugmentedLagrangianParameters() : num_iterations(12) {
-    lm_parameters = gtsam::LevenbergMarquardtParams();
-  }
+  AugmentedLagrangianParameters()
+      : Base(gtsam::LevenbergMarquardtParams()), num_iterations(12) {}
+
+  AugmentedLagrangianParameters(
+      const gtsam::LevenbergMarquardtParams& _lm_parameters)
+      : Base(_lm_parameters), num_iterations(12) {}
```
Comment (dellaert) [inline gtdynamics/optimizer/AugmentedLagrangianOptimizer.h:31]: Add argument with default 12

File: gtdynamics/optimizer/AugmentedLagrangianOptimizer.h [hunk 27]
```diff
@@ -20,11 +20,15 @@ namespace gtdynamics {
...
+  using Base = ConstrainedOptimizationParameters;
   size_t num_iterations;
 
-  AugmentedLagrangianParameters() : num_iterations(12) {
-    lm_parameters = gtsam::LevenbergMarquardtParams();
-  }
+  AugmentedLagrangianParameters()
+      : Base(gtsam::LevenbergMarquardtParams()), num_iterations(12) {}
+
+  AugmentedLagrangianParameters(
```
Comment (dellaert) [inline gtdynamics/optimizer/AugmentedLagrangianOptimizer.h:29]: Surprised no warning generated about explicit

File: gtdynamics/optimizer/Optimizer.h (hunk 13 above)
Comment (dellaert) [reply gtdynamics/optimizer/Optimizer.h:73]: Add TODO to remove then after discussing with team

File: gtdynamics/optimizer/PenaltyMethodOptimizer.h [hunk 28]
```diff
@@ -19,15 +19,23 @@ namespace gtdynamics {
...
   PenaltyMethodParameters()
-      : num_iterations(15), initial_mu(1.0), mu_increase_rate(2.0) {
-    lm_parameters = gtsam::LevenbergMarquardtParams();
-  }
+      : Base(gtsam::LevenbergMarquardtParams()),
+        num_iterations(15),
+        initial_mu(1.0),
+        mu_increase_rate(2.0) {}
+
+  PenaltyMethodParameters(const gtsam::LevenbergMarquardtParams& _lm_parameters)
```
Comment (dellaert) [inline gtdynamics/optimizer/PenaltyMethodOptimizer.h:34]: Add arguments with defaults

**Review by dellaert — APPROVED** [review]
Nice! Some small comments. Most important one is using emplace

File: gtdynamics/factors/PointGoalFactor.h (hunk 22 above)
Comment (yetongumich) [reply gtdynamics/factors/PointGoalFactor.h:100]: Done

File: gtdynamics/kinematics/KinematicsSlice.cpp (hunk 23 above)
Comment (yetongumich) [reply gtdynamics/kinematics/KinematicsSlice.cpp:79]: Done

File: gtdynamics/kinematics/KinematicsSlice.cpp (hunk 24 above)
Comment (yetongumich) [reply gtdynamics/kinematics/KinematicsSlice.cpp:96]: Done

File: gtdynamics/kinematics/KinematicsSlice.cpp (hunk 25 above)
Comment (yetongumich) [reply gtdynamics/kinematics/KinematicsSlice.cpp:63]: Done

File: gtdynamics/optimizer/AugmentedLagrangianOptimizer.h (hunk 26 above)
Comment (yetongumich) [reply gtdynamics/optimizer/AugmentedLagrangianOptimizer.h:31]: Done

File: gtdynamics/optimizer/AugmentedLagrangianOptimizer.h (hunk 27 above)
Comment (yetongumich) [reply gtdynamics/optimizer/AugmentedLagrangianOptimizer.h:29]: Done

File: gtdynamics/optimizer/PenaltyMethodOptimizer.h (hunk 28 above)
Comment (yetongumich) [reply gtdynamics/optimizer/PenaltyMethodOptimizer.h:34]: done
